content/blog/2025-03-27-valkey-8-1-0-rc1.md

```diff
@@ -0,0 +1,257 @@
+          v
+         ...
+```
+
+Regarding memory accesses, the above design requires at least 4 memory accesses to get the value in the case that the key is stored in the first `dictEntry` of the collision list. This might have a significant impact in cache locality when traversing the collision list search for the right key.
+
+The new hashtable implementation improves on the memory footprint required and on the cache locality by taking a different approach on the design of each bucket. Each bucket has a fixed size of 64 bytes (in a 64bit architecture), which usually corresponds to the size of a single cache line. The bucket structure stores 7 pointers, 8 bytes each, plus 8 bytes used for metadata. The 7 pointers are used to store up to 7 entries. Keys that map to the same bucket are all stored in the same bucket. If more than 7 keys map to the same bucket, a new bucket is linked to the previous one using the last of the 7 pointers. Each entry pointer, points to an object that may embed the key and the value.
+
+The 8 bytes of metadata is composed of 1 bit that indicates if there's child bucket linked, 7 bits that are used to indicate the presence of entries, and 7 bytes used as a secondary hash for each entry stored in the bucket.
+
+```
+       0          8       16      24      32      40      48      56      64  bytes
+       +------------------------------------------------------------------+
+       | Metadata | Entry | Entry | Entry | Entry | Entry | Entry | Entry |
+       +------------------------------------------------------------------+
+      /            ` - - . _ _
+     /                         `- - . _ _
+    /                                     ` - . _
+   +----------------------------------------------+
+   | c ppppppp hash hash hash hash hash hash hash |
+   +----------------------------------------------+
+    |    |       |
+    |    |      One byte of hash for each entry position in the bucket.
+    |    |
+    |   Presence bits. One bit for each entry position, indicating if an
+    |   entry present or not.
+    |
+   Chained? One bit. If set, the last entry is a child bucket pointer.
+```
+
+The secondary hashes are used for quickly eliminating hash collisions when looking up a key. In this way, we can avoid comparing the key for a mismatching key, except once in 256.
+
+This new design reduces by one the number of memory accesses required to lookup an entry when there are no collisions, but it reduces a lot more memory accesses to lookup an entry when there are key mapping collisions because each bucket can store 7 entries.
+
+It also reduces the memory footprint by a lot when there are collisions. With 3 keys mapping to the same bucket, the new design is already using less 8 bytes than the old design, which requires 72 bytes of memory (3 `dictEntry` objects). In the new design, the 64 byte bucket can store up to 7 entries, while the old design would require 168 bytes.
+
+
+### Other Performance Improvements
```

Comment:
Do we need this other section? Feels like we could just leave everything under performance.